HW0/P4/P4.py

```diff
@@ -65,12 +86,48 @@
 
     # B = ?
     # C = ?
+    A_matrix = propagation
+    B_diag = np.array([bx, by, bz, bvx, bvy, bvz])
+    B_matrix = np.diag(B_diag)
+    C_matrix = np.hstack([np.diag(np.array([rx, ry, rz])), np.zeros([3,3])])
 
     # Initial conditions for s0 and Sigma0
     # Compute the rest of sk using Eqs (2), (3), (4), and (5)
+    s2_matrix = np.zeros([6, K])
+    s2_matrix[:, 0] = np.array([0.0, 0.0, 2.0, 15.0, 3.5, 4.0]).transpose()
+    sig_diag = np.array([1, 1, 1, 1, 1, 1]) * 0.01
+    sig = np.diag(sig_diag)
+
+    def predictS():
+        s2_est = np.add(np.dot(propagation, s2_prev), acceleration)
+        return s2_est
+
+    def predictSig():
+        sig_est = np.linalg.inv(((A_matrix * sig) * A_matrix.transpose()) + B_matrix * B_matrix.transpose())
+        return sig_est
+
+    def updateSig():
+        sig = np.linalg.inv(sig_est + np.dot(C_matrix.transpose(), C_matrix))
+        return sig
+
+    def updateS():
+        s2 = np.dot(sig, (np.dot(sig_est, s2_est) + np.dot(C_matrix.transpose(), s_measure_trans[:, k])))
+        return s2
+
+    for k in range(1, K):
+        s2_prev = s2_matrix[:, k-1]
+        s2_est = predictS()
+        sig_est = predictSig()
+        sig = updateSig()
+        s2 = updateS()
+        s2_matrix[:, k] = s2
+
+    x_s2 = s2_matrix[0]
+    y_s2 = s2_matrix[1]
+    z_s2 = s2_matrix[2]
+
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-r', label='Filtered trajectory')
+    ax.plot(x_s2, y_s2, z_s2, '-r', label='Filtered trajectory')
 
     # Show the plot
     ax.legend()
```

Comment:
Everything else looks good. No errors except the matrix and dot product confusion.